### PR TITLE
Update design doc for Python Layer

### PR DESCRIPTION
**Pull request by reyoung**
This update is about how to implement layer function. The logic in this PR is as follow:

1. There are [around 150 lines of code](https://github.com/PaddlePaddle/Paddle/pull/4724) to implement `FullyConnected` layer if we do not provide any common functions. So it is necessary to give a mechanism for code reusing.
1. There is a comparison between global functions and a helper class. The disadvantages of global functions are 1) lack of a namespace. 2) Need to pass parameter every time.
   * There are two pieces of demo code for global functions and a helper class. It shows that a helper class can make code shorter and clearer.
1. How to implement a layer helper class.

### CODE REVIEW

File: doc/design/python_api.md [hunk 1]
```diff
@@ -179,41 +179,107 @@ init_attr={
...
+2. Create several decorators for layer function.
+3. Make layer classes as `functor`. The base class of layer has several member methods.
+
+Making several global functions is not appropriate since there are too many codes that can be reused and it is hard to organize and remember so many global functions. Making several decorators is not good either for the same reason.
+
+Making layer classes as `functor`, and giving a base class of all layer functor is the straightforward way for code reuse. The advantages are:
+
+1. The layer developer can figure out how many common functions they can use by just typing `self.` in an IDE.
+2. It is easy to understand and develop since inheritance is a common way to reuse code.
+3. It is transparent for end-users since we can wrap a functor class to a function.
```
Comment (jacquesqiao) [inline doc/design/python_api.md]: maybe Add a Conclusion part that we choose Functor to build our layer.

File: doc/design/python_api.md [hunk 2]
```diff
@@ -179,41 +179,107 @@ init_attr={
...
-                           inputs = None,
-                           outputs = [var],
-                           {column_name: column_name})
-    return var
+class Layer(object):
+    def __init__(self, **kwargs):  # accept any argument
+        self.kwargs = kwargs
+    
+    def __call__(self):
+        raise NotImplementedError("LayerBase is a ")
```
Comment (jacquesqiao) [inline doc/design/python_api.md]: complete the msg of `LayerBase is a `

File: doc/design/python_api.md [hunk 3]
```diff
@@ -179,41 +179,107 @@ init_attr={
...
+```
+
+Layer developers are free to use pre-defined methods that defined in `LayerBase`. Layer developers just need to overwrite the `__call__` method.
+
+### Export Functor Class as Function
+
+We can provide a decorator to wrap a functor class as a function. The code is as follows.
+
+```python
+def export(name):
```
Comment (jacquesqiao) [inline doc/design/python_api.md]: maybe `export_layer` or `layer`?  make user be sure what this decorator is used for.

File: doc/design/python_api.md [hunk 4]
```diff
@@ -179,40 +179,104 @@ init_attr={
...
+
+  pre_bias = helper.add_sum(mul_results)
+  pre_activation = helper.add_bias(pre_bias)
+  return helper.add_activation(pre_activation)
+```
+
+We not only use the fewer lines of code to write `fc_layer` but also make the code clearer to understand. At the same time, layer developers can figure out what function they can invoke by typing `helper.` in a python editor.
+
+
+### Implementation of layer helper
```
Comment (emailweixu) [inline doc/design/python_api.md]: Will the new api be compatible with v2 api?

**Pull request merged**